pyiceberg/utils/partition_map.py

```diff
@@ -0,0 +1,45 @@
+from typing import Any, Callable, Dict, Generic, Iterator, Optional, Tuple, TypeVar
+
+from pyiceberg.partitioning import PartitionSpec
+from pyiceberg.typedef import Record
+
+T = TypeVar("T")
+
+
+class PartitionMap(Generic[T]):
+    def __init__(self, specs_by_id: Optional[Dict[int, PartitionSpec]] = None) -> None:
+        self._specs_by_id = specs_by_id or {}
+        self._map: Dict[Tuple[int, Tuple[Any, ...]], T] = {}
+
+    def get(self, spec_id: int, partition: Optional[Record]) -> Optional[T]:
+        """Get a value by spec ID and partition."""
+        key = self._make_key(spec_id, partition)
+        return self._map.get(key)
+
+    def put(self, spec_id: int, partition: Optional[Record], value: T) -> None:
+        """Put a value by spec ID and partition."""
+        key = self._make_key(spec_id, partition)
+        self._map[key] = value
+
+    def compute_if_absent(self, spec_id: int, partition: Optional[Record], factory: Callable[[], T]) -> T:
+        """Get a value by spec ID and partition, creating it if it doesn't exist."""
```

Comment:
Should we add a validation here maybe for put too but I don't see it used.

```suggestion
        """Get a value by spec ID and partition, creating it if it doesn't exist."""
        if spec_id not in self._specs:
            raise ValueError(f"Cannot find spec with ID {spec_id}: self._specs")
```

Reply:
Will add the validation. I added the put method to maintain the API parity with the Java implementation hoping that it will be used in the future. I can remove it as it is not being currently used.